kernel/kernel-api/src/test/scala/io/delta/kernel/types/FieldMetadataSuite.scala

```diff
@@ -0,0 +1,83 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.types
+
+import org.assertj.core.api.Assertions.{assertThat, assertThatThrownBy}
+import org.scalatest.funsuite.AnyFunSuite
+
+class FieldMetadataSuite extends AnyFunSuite {
+
+  test("retrieving non-existing key returns null") {
+    assertThat(FieldMetadata.builder().build().get("non-existing")).isNull()
+    assertThat(FieldMetadata.builder().putBoolean("key", false).build()
+      .get("non-existing", Long.getClass)).isNull()
+  }
+
+  test("retrieving key with null value should never throw") {
+    val meta = FieldMetadata.builder().putNull("nullKey").build()
+    assertThat(meta.get("nullKey", classOf[Integer])).isNull()
+    assertThat(meta.get("nullKey", classOf[java.lang.Boolean])).isNull()
+  }
+
+  test("retrieving key with wrong type throws exception") {
+    val meta = FieldMetadata.builder()
+      .putLong("longKey", 23L)
+      .putDouble("doubleKey", 23.0)
+      .putBoolean("booleanKey", true)
+      .putString("stringKey", "random")
+      .putFieldMetadataArray("fieldMetadataArrayKey", Seq(FieldMetadata.builder().build()).toArray)
+      .putFieldMetadata("fieldMetadataKey", FieldMetadata.builder().build())
+      .build
+
+    assertThatThrownBy(() => meta.get("longKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+
+    assertThatThrownBy(() => meta.get("doubleKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+
+    assertThatThrownBy(() => meta.get("booleanKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+
+    assertThatThrownBy(() => meta.get("stringKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+
+    assertThatThrownBy(() => meta.get("fieldMetadataArrayKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+
+    assertThatThrownBy(() => meta.get("fieldMetadataKey", classOf[Integer]))
+      .isInstanceOf(classOf[IllegalArgumentException])
+  }
+
+  test("retrieving key with correct type returns value") {
+    val innerMeta = FieldMetadata.builder().putBoolean("key", true).build()
+    val meta = FieldMetadata.builder()
+      .putLong("longKey", 23L)
+      .putDouble("doubleKey", 23.0)
+      .putBoolean("booleanKey", true)
+      .putString("stringKey", "random")
+      .putFieldMetadataArray("fieldMetadataArrayKey", Seq(innerMeta).toArray)
+      .putFieldMetadata("fieldMetadataKey", innerMeta)
+      .build
+
+    assertThat(meta.get("longKey", classOf[java.lang.Long])).isEqualTo(23L)
```

Comment:
one qn: what happens when you try to access the value as `java.lang.Integer`?

Reply:
that would fail